### PR TITLE
feature/eng-2509-add-earliness-and-lateness-penalty

### PR DESCRIPTION
**Pull request by nilsbeck**
# Description

This PR adds the Earliness and Lateness objective to the model. It will measure how early or late each stop are respective to their target time. This deviation will be multiplied by a factor which can be specified in the input. If no factor is given, the default factor is 1.

## Changes

- adds a new earliness objective. 
- extends the `LatestStart` and `LatestEnd` constructs to receive a factor.

### CODE REVIEW

File: nextroute/model_latestend.go [hunk 1]
```diff
@@ -18,6 +18,11 @@ type LatestEnd interface {
 
 	// Lateness returns the lateness of a stop. The lateness is the difference
```
Comment (davidrijsman) [inline nextroute/model_latestend.go:19]: Doc sentence seems to be cutoff

File: nextroute/model_latestend.go [hunk 2]
```diff
@@ -18,6 +18,11 @@ type LatestEnd interface {
...
 	Lateness(stop SolutionStop) float64
+
+	// SetLatenessFactor adds a factor with which a deviating stop is
+	// multiplied. This is only taken into account if the construct is used as
+	// an objective.
+	SetLatenessFactor(factor StopExpression)
```
Comment (davidrijsman) [inline nextroute/model_latestend.go]: I would change this to such that you can check factor >= 0
`SetFactor(factor float64, stop StopExpression) ` 
and also introduce a so one can always verify what has been used:
`Factor(stop StopExpression) float64`

File: nextroute/model_lateststart.go [hunk 3]
```diff
@@ -18,6 +18,11 @@ type LatestStart interface {
 
 	// Lateness returns the lateness of a stop. The lateness is the difference
```
Comment (davidrijsman) [inline nextroute/model_lateststart.go:19]: Doc seems to be cutoff (no do either)

File: nextroute/model_lateststart.go [hunk 4]
```diff
@@ -18,6 +18,11 @@ type LatestStart interface {
...
 	Lateness(stop SolutionStop) float64
+
+	// SetLatenessFactor adds a factor with which a deviating stop is
+	// multiplied. This is only taken into account if the construct is used as
+	// an objective.
+	SetLatenessFactor(factor StopExpression)
```
Comment (davidrijsman) [inline nextroute/model_lateststart.go]: See model_latestend for how I suggest we do it

**Review by davidrijsman — CHANGES_REQUESTED** [review]

File: nextroute/model_latestend.go (hunk 1 above)
Comment (nilsbeck) [reply nextroute/model_latestend.go:19]: fixed

File: nextroute/model_latestend.go (hunk 2 above)
Comment (nilsbeck) [reply nextroute/model_latestend.go]: changed

File: nextroute/model_lateststart.go (hunk 3 above)
Comment (nilsbeck) [reply nextroute/model_lateststart.go:19]: fixed

File: nextroute/model_lateststart.go (hunk 4 above)
Comment (nilsbeck) [reply nextroute/model_lateststart.go]: changed

**Review by davidrijsman — DISMISSED** [review]

**Review by larsbeck — DISMISSED** [review]

**Review by davidrijsman — APPROVED** [review]

**Pull request merged**